File: Phase2/src/main/java/controller/QueryHandler.java
package controller;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import model.DocumentFile;
import model.InvertedIndex;
import model.query.ANDOperator;
import model.query.NOTOperator;
import model.query.OROperator;
import model.query.Operator;
import util.WordUtil;

public class QueryHandler {
    private final InvertedIndex invertedIndex;
    private final List<Operator> queriesList;
    private final char orSymbol = '+';
    private final char notSymbol = '-';

    public QueryHandler(InvertedIndex invertedIndex) {
        this.invertedIndex = invertedIndex;
        queriesList = new ArrayList<>();
    }

    public Set<DocumentFile> search(String line, WordUtil wordUtil) {
        setQueryList(line);
        Set<DocumentFile> resultSet = new HashSet<>(invertedIndex.getDocuments());
        for (Operator query : queriesList)
            resultSet = query.operate(resultSet, wordUtil);
        return resultSet;
    }

    private void setQueryList(String query) {
        String[] subQueries = query.split("\\s+");
        for (String subQuery : subQueries)
            queriesList.add(getNewInstance(subQuery, invertedIndex));
    }

    private Operator getNewInstance(String queryString, InvertedIndex index){
        if(queryString.charAt(0) == orSymbol)
            return new OROperator(queryString, index);
        else if(queryString.charAt(0) == notSymbol)
            return new NOTOperator(queryString, index);
        return new ANDOperator(queryString, index);
    }
}

File: Phase5/SearchEngine/SearchEngine/Controller/QueryHandler.cs
using SearchEngine.Model;
using System;
using System.Collections.Generic;
using System.Text;
using SearchEngine.Model.Operator;

namespace SearchEngine.Controller
{
    public class QueryHandler : IQueryHandler
    {
        private List<IOperator> _operatorsList;
        public QueryHandler(string searchQuery)
        {
            this._operatorsList = new List<IOperator>();
            MakeOperatorsList(searchQuery.Split(" "));
        }

        public List<IDocument> OperateOnQuery(IInvertedIndex index)
        {
            var result = new List<IDocument>();
            foreach (IOperator op in _operatorsList)
                result = op.Operate(index, result);
            return result;
        }

        private void MakeOperatorsList(string[] searchQuery)
        {
            foreach (string query in searchQuery)
            {
                if (query[0] == '+')
                    _operatorsList.Add(new OrOperator(query.Substring(1)));
                else if (query[0] == '-')
                    _operatorsList.Add(new NotOperator(query.Substring(1)));
                else
                    _operatorsList.Add(new AndOperator(query));
            }
        }
    }
}
